File: kythe/java/com/google/devtools/kythe/util/KytheURI.java
/*
 * Copyright 2014 The Kythe Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.devtools.kythe.util;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Strings.isNullOrEmpty;
import static com.google.common.base.Strings.nullToEmpty;

import com.google.common.base.Joiner;
import com.google.common.base.Splitter;
import com.google.common.escape.Escaper;
import com.google.common.net.PercentEscaper;
import com.google.devtools.kythe.proto.CorpusPath;
import com.google.devtools.kythe.proto.Storage.VName;
import java.io.Serializable;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.Map;
import org.checkerframework.checker.nullness.qual.Nullable;

/**
 * /{@link String} realization of a {@link VName}.
 *
 * <p>Specification: //kythe/util/go/kytheuri/kythe-uri-spec.txt
 */
public class KytheURI implements Serializable {
  private static final long serialVersionUID = 2726281203803801095L;

  public static final String SCHEME_LABEL = "kythe";
  public static final String SCHEME = SCHEME_LABEL + ":";
  public static final KytheURI EMPTY = new KytheURI();

  private static final String SAFE_CHARS = ".-_~";
  private static final Escaper PATH_ESCAPER = new PercentEscaper(SAFE_CHARS + "/", false);
  private static final Escaper ALL_ESCAPER = new PercentEscaper(SAFE_CHARS, false);
  private static final Splitter SIGNATURE_SPLITTER = Splitter.on('#');
  private static final Splitter.MapSplitter PARAMS_SPLITTER =
      Splitter.on('?').withKeyValueSeparator("=");
  private static final Splitter PATH_SPLITTER = Splitter.on('/');

  private final VName vName;

  /** Construct a new {@link KytheURI}. */
  public KytheURI(String signature, String corpus, String root, String path, String language) {
    this(
        VName.newBuilder()
            .setSignature(nullToEmpty(signature))
            .setCorpus(nullToEmpty(corpus))
            .setRoot(nullToEmpty(root))
            .setPath(nullToEmpty(path))
            .setLanguage(nullToEmpty(language))
            .build());
  }

  /** Constructs an empty {@link KytheURI}. */
  private KytheURI() {
    this(VName.getDefaultInstance());
  }

  /** Unpacks a {@link VName} into a new {@link KytheURI}. */
  public KytheURI(VName vName) {
    this.vName = vName;
  }

  /** Unpacks a {@link CorpusPath} into a new {@link KytheURI}. */
  public KytheURI(CorpusPath cp) {
    this(
        VName.newBuilder()
            .setCorpus(cp.getCorpus())
            .setRoot(cp.getRoot())
            .setPath(cp.getPath())
            .build());
  }

  /** Returns the {@link KytheURI}'s signature. */
  public String getSignature() {
    return vName.getSignature();
  }

  /** Returns the {@link KytheURI}'s corpus name. */
  public String getCorpus() {
    return vName.getCorpus();
  }

  /** Returns the {@link KytheURI}'s corpus path. */
  public String getPath() {
    return vName.getPath();
  }

  /** Returns the {@link KytheURI}'s corpus root. */
  public String getRoot() {
    return vName.getRoot();
  }

  /** Returns the {@link KytheURI}'s language. */
  public String getLanguage() {
    return vName.getLanguage();
  }

  /** Returns an equivalent {@link VName}. */
  public VName toVName() {
    return vName;
  }

  /** Returns the {@link CorpusPath} components of the uri. */
  public CorpusPath toCorpusPath() {
    return CorpusPath.newBuilder()
        .setCorpus(vName.getCorpus())
        .setRoot(vName.getRoot())
        .setPath(vName.getPath())
        .build();
  }

  @Override
  public String toString() {
    StringBuilder b =
        new StringBuilder(
            SCHEME.length()
                + "//?=?=?=#".length() // assume all punctuation is necessary
                + getCorpus().length()
                + getPath().length()
                + getRoot().length()
                + getLanguage().length()
                + getSignature().length());
    b.append(SCHEME);
    if (!vName.getCorpus().isEmpty()) {
      b.append("//");
      b.append(PATH_ESCAPER.escape(vName.getCorpus()));
    }
    attr(b, "lang", ALL_ESCAPER.escape(vName.getLanguage()));
    attr(b, "path", PATH_ESCAPER.escape(cleanPath(vName.getPath())));
    attr(b, "root", PATH_ESCAPER.escape(vName.getRoot()));
    if (!vName.getSignature().isEmpty()) {
      b.append("#");
      b.append(ALL_ESCAPER.escape(vName.getSignature()));
    }
    return b.toString();
  }

  @Override
  public int hashCode() {
    return vName.hashCode();
  }

  @Override
  public boolean equals(Object o) {
    return this == o || (o instanceof KytheURI && vName.equals(((KytheURI) o).vName));
  }

  /** Returns an equivalent Kythe ticket for the given {@link VName}. */
  public static String asString(VName vName) {
    return new KytheURI(vName).toString();
  }

  /** Returns an equivalent Kythe ticket for the given {@link CorpusPath}. */
  public static String asString(CorpusPath cp) {
    return new KytheURI(cp).toString();
  }

  /** Parses the given string to produce a new {@link KytheURI}. */
  public static KytheURI parse(String str) {
    checkNotNull(str, "str must be non-null");
    if (str.isEmpty() || str.equals(SCHEME) || str.equals(SCHEME + "//")) {
      return EMPTY;
    }
    String original = str;

    // Check for a scheme label.  This may be empty; but if present, it must be
    // our expected scheme.
    if (str.startsWith(SCHEME)) {
      str = str.substring(SCHEME.length());
    }

    String signature = null;
    String corpus = null;
    String path = null;
    String root = null;
    String lang = null;

    // Split corpus/attributes from signature.
    Iterator<String> parts = SIGNATURE_SPLITTER.split(str).iterator();
    String head = parts.next();
    if (parts.hasNext()) {
      signature = parts.next();
    }
    checkArgument(!parts.hasNext(), "URI has multiple fragments: %s", original);

    // Remove corpus prefix from attributes.
    int firstParam = head.indexOf('?');
    firstParam = firstParam < 0 ? head.length() : firstParam;
    if (head.startsWith("//")) {
      corpus = head.substring(2, firstParam);
      head = head.substring(firstParam);
    } else {
      checkArgument(firstParam == 0, "invalid URI scheme: %s", original);
    }

    // If there are any attributes, parse them.  We allow valid attributes to
    // occur in any order, even if it is not canonical.
    if (!head.isEmpty()) {
      Map<String, String> params = PARAMS_SPLITTER.split(head.substring(1));

      for (Map.Entry<String, String> e : params.entrySet()) {
        switch (e.getKey()) {
          case "path":
            path = e.getValue();
            break;
          case "root":
            root = e.getValue();
            break;
          case "lang":
            lang = e.getValue();
            break;
          default:
            checkArgument(false, "invalid attribute: %s (value: %s)", e.getKey(), e.getValue());
            break;
        }
      }
    }

    return new KytheURI(
        decode(signature), decode(corpus), decode(root), decode(path), decode(lang));
  }

  /** Parses the given Kythe ticket string to produce a new {@link VName}. */
  public static VName parseVName(String ticket) {
    return parse(ticket).toVName();
  }

  private static @Nullable String decode(String str) {
    if (isNullOrEmpty(str)) {
      return null;
    }
    try {
      return URLDecoder.decode(str.replace("+", "%2B"), "UTF-8");
    } catch (java.io.UnsupportedEncodingException e) {
      throw new RuntimeException(e);
    }
  }

  /** Returns a new {@link KytheURI.Builder}. */
  public static Builder newBuilder() {
    return new Builder();
  }

  public static class Builder {
    private final VName.Builder builder = VName.newBuilder();

    public Builder setSignature(String signature) {
      builder.setSignature(nullToEmpty(signature));
      return this;
    }

    public Builder setCorpus(String corpus) {
      builder.setCorpus(nullToEmpty(corpus));
      return this;
    }

    public Builder setPath(String path) {
      builder.setPath(nullToEmpty(path));
      return this;
    }

    public Builder setRoot(String root) {
      builder.setRoot(nullToEmpty(root));
      return this;
    }

    public Builder setLanguage(String language) {
      builder.setLanguage(nullToEmpty(language));
      return this;
    }

    public KytheURI build() {
      return new KytheURI(builder.build());
    }
  }

  private static void attr(StringBuilder b, String name, String value) {
    if (value.isEmpty()) {
      return;
    }
    b.append("?");
    b.append(name);
    b.append("=");
    b.append(value);
  }

  /**
   * Returns a lexically cleaned path, with repeated slashes and "." path components removed, and
   * ".." path components rewound as far as possible without touching the filesystem.
   */
  private static String cleanPath(String path) {
    ArrayList<String> clean = new ArrayList<>();
    if (!path.isEmpty() && path.charAt(0) == '/') {
      clean.add("");
    }
    for (String part : PATH_SPLITTER.split(path)) {
      if (part.isEmpty() || part.equals(".")) {
        continue; // skip empty path components and "here" markers.
      } else if (part.equals("..") && !clean.isEmpty()) {
        clean.remove(clean.size() - 1);
        continue; // back off if possible for "up" (..) markers.
      } else {
        clean.add(part);
      }
    }
    return Joiner.on('/').join(clean);
  }
}

File: kythe/cxx/common/kythe_uri.cc
/*
 * Copyright 2015 The Kythe Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "kythe/cxx/common/kythe_uri.h"

#include <cstddef>
#include <string>
#include <utility>

#include "absl/strings/match.h"
#include "absl/strings/str_split.h"
#include "absl/strings/string_view.h"
#include "kythe/cxx/common/path_utils.h"
#include "kythe/proto/storage.pb.h"

namespace kythe {
namespace {

constexpr char kHexDigits[] = "0123456789ABCDEF";
/// The URI scheme label for Kythe.
constexpr char kUriScheme[] = "kythe";
constexpr char kUriPrefix[] = "kythe:";

/// \brief Returns whether a byte should be escaped.
/// \param mode The escaping mode to use.
/// \param c The byte to examine.
bool should_escape(UriEscapeMode mode, char c) {
  return !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
           (c >= '0' && c <= '9') || c == '-' || c == '.' || c == '_' ||
           c == '~' || (mode == UriEscapeMode::kEscapePaths && c == '/'));
}

/// \brief Returns the value of a hex digit.
/// \param digit The hex digit.
/// \return The value of the digit, or -1 if it was not a hex digit.
int value_for_hex_digit(char c) {
  if (c >= '0' && c <= '9') {
    return c - '0';
  } else if (c >= 'a' && c <= 'f') {
    return 10 + (c - 'a');
  } else if (c >= 'A' && c <= 'F') {
    return 10 + (c - 'A');
  }
  return -1;
}

std::pair<absl::string_view, absl::string_view> Split(absl::string_view input,
                                                      char ch) {
  return absl::StrSplit(input, absl::MaxSplits(ch, 1));
}

}  // namespace

std::string UriEscape(UriEscapeMode mode, absl::string_view uri) {
  size_t num_escapes = 0;
  for (char c : uri) {
    if (should_escape(mode, c)) {
      ++num_escapes;
    }
  }
  std::string result;
  result.reserve(num_escapes * 2 + uri.size());
  for (char c : uri) {
    if (should_escape(mode, c)) {
      result.push_back('%');
      result.push_back(kHexDigits[(c >> 4) & 0xF]);
      result.push_back(kHexDigits[c & 0xF]);
    } else {
      result.push_back(c);
    }
  }
  return result;
}

/// \brief URI-unescapes a string.
/// \param string The string to unescape.
/// \return A pair of (success, error-or-unescaped-string).
std::pair<bool, std::string> UriUnescape(absl::string_view string) {
  size_t num_escapes = 0;
  for (size_t i = 0, s = string.size(); i < s; ++i) {
    if (string[i] == '%') {
      ++num_escapes;
      if (i + 3 > string.size()) {
        return std::make_pair(false, "bad escape");
      }
      i += 2;
    }
  }
  std::string result;
  result.reserve(string.size() - num_escapes * 2);
  for (size_t i = 0, s = string.size(); i < s; ++i) {
    char c = string[i];
    if (c != '%') {
      result.push_back(c);
      continue;
    }
    int high = value_for_hex_digit(string[++i]);
    int low = value_for_hex_digit(string[++i]);
    if (high < 0 || low < 0) {
      return std::make_pair(false, "bad hex digit");
    }
    result.push_back((high << 4) | low);
  }
  return std::make_pair(true, result);
}

std::string URI::ToString() const {
  std::string result = kUriPrefix;
  if (vname_.signature().empty() && vname_.path().empty() &&
      vname_.root().empty() && vname_.corpus().empty() &&
      vname_.language().empty()) {
    return result;
  }
  absl::string_view signature = vname_.signature();
  absl::string_view path = vname_.path();
  absl::string_view corpus = vname_.corpus();
  absl::string_view language = vname_.language();
  absl::string_view root = vname_.root();
  if (!corpus.empty()) {
    result.append("//");
    result.append(UriEscape(UriEscapeMode::kEscapePaths, corpus));
  }
  if (!language.empty()) {
    result.append("?lang=");
    result.append(UriEscape(UriEscapeMode::kEscapeAll, language));
  }
  if (!path.empty()) {
    result.append("?path=");
    result.append(UriEscape(UriEscapeMode::kEscapePaths, CleanPath(path)));
  }
  if (!root.empty()) {
    result.append("?root=");
    result.append(UriEscape(UriEscapeMode::kEscapePaths, root));
  }
  if (!signature.empty()) {
    result.push_back('#');
    result.append(UriEscape(UriEscapeMode::kEscapeAll, signature));
  }
  return result;
}

/// \brief Separate out the scheme component of `uri` if one exists.
/// \return (scheme, tail) if there was a scheme; ("", uri) otherwise.
static std::pair<absl::string_view, absl::string_view> SplitScheme(
    absl::string_view uri) {
  for (size_t i = 0, s = uri.size(); i != s; ++i) {
    char c = uri[i];
    if (c == ':') {
      return std::make_pair(uri.substr(0, i), uri.substr(i));
    } else if ((i == 0 &&
                ((c >= '0' && c <= '9') || c == '+' || c == '-' || c == '.')) ||
               !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'))) {
      break;
    }
  }
  return std::make_pair(absl::string_view(), uri);
}

URI::URI(const kythe::proto::VName& from_vname) : vname_(from_vname) {}

bool URI::ParseString(absl::string_view uri) {
  auto head_fragment = Split(uri, '#');
  auto head = head_fragment.first, fragment = head_fragment.second;
  auto scheme_head = SplitScheme(head);
  auto scheme = scheme_head.first;
  head = scheme_head.second;
  if (scheme.empty()) {
    if (absl::StartsWith(head, ":")) {
      return false;
    }
  } else if (scheme != kUriScheme) {
    return false;
  } else if (!head.empty()) {
    head.remove_prefix(1);
  }
  auto head_attrs = Split(head, '?');
  head = head_attrs.first;
  auto attrs = head_attrs.second;
  std::string corpus;
  if (!head.empty()) {
    if (!absl::StartsWith(head, "//")) {
      return false;
    }
    auto maybe_corpus = UriUnescape(head.substr(2));
    if (!maybe_corpus.first) {
      return false;
    }
    corpus = maybe_corpus.second;
  }
  auto maybe_sig = UriUnescape(fragment);
  if (!maybe_sig.first) {
    return false;
  }
  auto signature = maybe_sig.second;
  while (!attrs.empty()) {
    auto attr_rest = Split(attrs, '?');
    auto attr = attr_rest.first;
    attrs = attr_rest.second;
    auto name_value = Split(attr, '=');
    auto maybe_value = UriUnescape(name_value.second);
    if (!maybe_value.first || maybe_value.second.empty()) {
      return false;
    }
    if (name_value.first == "lang") {
      vname_.set_language(maybe_value.second);
    } else if (name_value.first == "root") {
      vname_.set_root(maybe_value.second);
    } else if (name_value.first == "path") {
      vname_.set_path(CleanPath(maybe_value.second));
    } else {
      return false;
    }
  }
  vname_.set_signature(signature);
  vname_.set_corpus(corpus);
  return true;
}

}  // namespace kythe
